lib/vagrant/action/builtin/cloud_init_wait.rb

```diff
@@ -0,0 +1,26 @@
+module Vagrant
+  module Action
+    module Builtin
+      class CloudInitWait
+
+        def initialize(app, env)
+          @app    = app
+          @logger = Log4r::Logger.new("vagrant::action::builtin::cloudinitwait")
+        end
+
+        def call(env)
+          machine = env[:machine]
+          if !machine.config.vm.cloud_init_configs.empty?
+            if machine.communicate.test("command -v cloud-init")
+              @logger.info("Waiting for cloud init")
+              machine.communicate.sudo("cloud-init status --wait")
```

Comment:
It might be nice to pop up a message that says that cloud-init has finished. We should also check the result of this command and raise a cloud-init specific error, rather than the generic SSH error:


```suggestion
              result = machine.communicate.sudo("cloud-init status --wait", error_check: false)
```
(etc.)